libs/NfcCoreLib/lib/LibNfc/phLibNfc_Hci.c

```diff
@@ -2150,29 +2169,59 @@ NFCSTATUS phLibNfc_CreateNextApduPipe(pphLibNfc_Context_t pLibContext)
         if (seInfo->eSE_Type != phLibNfc_SE_Type_eSE ||
             seHciVersion < phHciNfc_e_HciVersion12)
         {
-            PH_LOG_LIBNFC_INFO_STR("SE with index %u does not need APDU pipe", seIndex);
+            PH_LOG_LIBNFC_INFO_STR("SE with index %u hostId=0x%02X does not need APDU pipe", seIndex, seInfo->hciHostId);
             continue;
         }
+        // Check if APDU pipe is already present for SE.
+        if (phLibNfc_IsAPDUPipePresent(seInfo->hciHostId, pLibContext))
+        {
+            PH_LOG_LIBNFC_INFO_STR("SE with index %u hostId=0x%02X APDU pipe is already present!!", seIndex, seInfo->hciHostId);
+            continue;
+        }
 
         PHLIBNFC_INIT_SEQUENCE(pLibContext, gphLibNfc_HciChildDevCreateApduPipeInitSequence);
         wStatus = phLibNfc_SeqHandler(pLibContext, NFCSTATUS_SUCCESS, NULL);
 
         if (wStatus != NFCSTATUS_PENDING)
         {
-            PH_LOG_LIBNFC_CRIT_X32MSG("Failed to initiate APDU pipe creation, error", wStatus);
+            PH_LOG_LIBNFC_CRIT_X32MSG("Failed to initiate APDU pipe creation , error", wStatus);
             continue;
         }
 
         break;
     }
 
+
     if (wStatus != NFCSTATUS_PENDING)
     {
         PH_LOG_LIBNFC_INFO_STR("APDU pipes initialization is finished.");
-        PHLIBNFC_INIT_SEQUENCE(pLibContext, gphLibNfc_HciEndInitSequence);
-        wStatus = phLibNfc_SeqHandler(pLibContext, NFCSTATUS_SUCCESS, NULL);
+        //PHLIBNFC_INIT_SEQUENCE(pLibContext, gphLibNfc_HciEndInitSequence);
+        //wStatus = phLibNfc_SeqHandler(pLibContext, NFCSTATUS_SUCCESS, NULL);
+        wStatus = phLibNfc_LaunchNfceeDiscCompleteSequence(pLibContext, NFCSTATUS_SUCCESS, NULL);
     }
 
     PH_LOG_LIBNFC_FUNC_EXIT();
     return wStatus;
 }
+static BOOL phLibNfc_IsAPDUPipePresent(uint8_t hostId, void *pContext)
+{
+    pphLibNfc_LibContext_t pLibContext = (pphLibNfc_LibContext_t)pContext;
+    pphHciNfc_HciContext_t pHciContext;
+    BOOL status = TRUE;
+    PH_LOG_LIBNFC_FUNC_ENTRY();
+    if ((NULL != pLibContext) && (phLibNfc_GetContext() == pLibContext))
+    {
+        pHciContext = pLibContext->pHciContext;
+        if ((hostId >= phHciNfc_e_ProprietaryHostID_Min) && (hostId <= phHciNfc_e_ProprietaryHostID_Max))
+        {
+            PH_LOG_LIBNFC_CRIT_STR("Pipe Id present = %d", pHciContext->aSEPipeList[PHHCI_ESE_APDU_PIPE_LIST_INDEX].bPipeId);
+            if (pHciContext->aSEPipeList[PHHCI_ESE_APDU_PIPE_LIST_INDEX].bPipeId == PHHCINFC_NO_PIPE_DATA ||
+                pHciContext->aSEPipeList[PHHCI_ESE_APDU_PIPE_LIST_INDEX].bPipeId == 0x00)
+            {
+                return FALSE;
+            }
+        }
+    }
+    PH_LOG_LIBNFC_FUNC_EXIT();
+    return status;
+}
```

Comment:
This function returns TRUE in case of invalid parameters, or `hostId` being outside of proprietary range. If this logic is on purpose, please add a comment with your reasoning.

Reply:
The negative logic is on purpose.
The Secure Element pipe creation will be done if the function returns  FALSE.
Will add comments.

Reply:
We do not necessarily want to create APDU pipe for every `hostId` which is not in the proprietary range or is not an eSE. Supported ETSI version also matters in this context.

And in general, name of the function states `IsAPDUPipePresent` which implies that if it returns TRUE that APDU pipe _has been created_ rather than _needs to initiate APDU pipe creation_.

Reply:
If the function IsAPDUPipePresent() returns FALSE, then only there is a need to initiate APDU pipe creation.

Like IsAPDUPipePresent() returns TRUE ---> No need for APDU pipe creation.
And if IsAPDUPipePresent() returns FALSE ---> Go ahead for APDU pipe creation.

Check caller @ line 2176..
The ETSI check is done before this check.

Reply:
Based on the name of the function, code readers would think:
IsAPDUPipePresent() returns TRUE ---> APDU pipe is present (already exists).
IsAPDUPipePresent() returns FALSE ---> APDU pipe for this host is not present.

In the TRUE case it is not always a case.

Perhaps this misinterpretation could be solved as easy as renaming the function to something like `IsAPDUPipeCreationNeeded()`

Reply:
Change is funcion name needs rearranging of code also.
Even if the Host Id is not valid function returns TRUE , again no need to create the Pipe.

Here the caller 
           if(  IsAPDUPipePresent())
          {
             //do nothing
          }
         else
        {
         //Create Pipe
        }